iroh-net/src/derp.rs

```diff
@@ -184,180 +186,116 @@ impl From<FrameType> for u8 {
-async fn write_frame<'a>(
-    mut writer: impl AsyncWrite + Unpin,
-    frame_type: FrameType,
-    bytes: &[&[u8]],
-) -> Result<()> {
-    let bytes_len: usize = bytes.iter().map(|b| b.len()).sum();
-    ensure!(
-        bytes_len <= MAX_FRAME_SIZE,
-        "unreasonably large frame write"
-    );
-    write_frame_header(&mut writer, frame_type, bytes_len).await?;
-    for b in bytes {
-        writer.write_all(b).await?;
-    }
-    Ok(())
-}
-
 /// AsyncWrites a complete frame, errors if it is unable to write within the given `timeout`.
 /// Ignores the timeout if `timeout.is_zero()`
 ///
 /// Does not flush.
-async fn write_frame_timeout(
-    writer: impl AsyncWrite + Unpin,
-    frame_type: FrameType,
-    bytes: &[&[u8]],
+async fn write_frame_timeout<S: Sink<WriteFrame, Error = std::io::Error> + Unpin>(
+    mut writer: S,
+    frame: WriteFrame,
     timeout: Option<Duration>,
 ) -> Result<()> {
     if let Some(duration) = timeout {
-        tokio::time::timeout(duration, write_frame(writer, frame_type, bytes)).await??;
-        Ok(())
+        tokio::time::timeout(duration, writer.send(frame)).await??;
     } else {
-        write_frame(writer, frame_type, bytes).await
+        writer.send(frame).await?;
     }
+
+    Ok(())
 }
 
 /// Writes a `FrameType::ClientInfo`, including the client's [`PublicKey`],
 /// and the client's [`ClientInfo`], sealed using the server's [`PublicKey`].
 ///
 /// Flushes after writing.
-pub(crate) async fn send_client_key<W: AsyncWrite + Unpin>(
-    mut writer: W,
+pub(crate) async fn send_client_key<S: Sink<WriteFrame, Error = std::io::Error> + Unpin>(
+    mut writer: S,
     shared_secret: &SharedSecret,
     client_public_key: &PublicKey,
     client_info: &ClientInfo,
 ) -> Result<()> {
     let mut msg = postcard::to_stdvec(client_info)?;
     shared_secret.seal(&mut msg);
-    write_frame(
-        &mut writer,
-        FrameType::ClientInfo,
-        &[client_public_key.as_bytes(), &msg],
-    )
-    .await?;
+    writer
+        .send(WriteFrame::ClientInfo {
+            client_public_key: *client_public_key,
+            encrypted_message: msg.into(),
+        })
+        .await?;
     writer.flush().await?;
     Ok(())
 }
 
 /// Reads the `FrameType::ClientInfo` frame from the client (its proof of identity)
 /// upon it's initial connection.
-async fn recv_client_key<R: AsyncRead + Unpin>(
+async fn recv_client_key<S: Stream<Item = anyhow::Result<WriteFrame>> + Unpin>(
     secret_key: SecretKey,
-    mut reader: R,
+    stream: S,
 ) -> Result<(PublicKey, ClientInfo, SharedSecret)> {
-    let mut buf = BytesMut::new();
     // the client is untrusted at this point, limit the input size even smaller than our usual
     // maximum frame size, and give a timeout
-    let (frame_type, _) = read_frame_timeout(
-        &mut reader,
-        256 * 1024,
-        &mut buf,
-        Some(Duration::from_secs(10)),
+
+    // TODO: variable recv size: 256 * 1024
```

Comment:
I don't think there is a way to do this with the tokio_codec framework...

Reply:
there is, we need to set this configurable on the derpcodec struct, and then set it and set it back